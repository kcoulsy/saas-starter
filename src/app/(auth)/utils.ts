import { pageRoutes } from '@src/constants/routes';
import { authOptions } from '@src/pages/api/auth/[...nextauth]';
import { getServerSession } from 'next-auth';
import { redirect } from 'next/navigation';

export async function redirectIfLoggedIn() {
  const session = await getServerSession(authOptions);

  if (session?.user?.email) {
    redirect(pageRoutes.dashboard);
  }
}
