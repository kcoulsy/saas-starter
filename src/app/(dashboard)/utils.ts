import { getServerSession } from 'next-auth';
import { redirect } from 'next/navigation';
import { pageRoutes } from '@src/constants/routes';
import { authOptions } from '@src/server/auth';

export async function redirectIfNotLoggedIn() {
  const session = await getServerSession(authOptions);

  if (!session?.user?.email) {
    redirect(pageRoutes.login);
  }
}
