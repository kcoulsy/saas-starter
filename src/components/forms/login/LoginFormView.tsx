import React, { useState } from 'react';
import type { UseFormRegisterReturn } from 'react-hook-form';

interface Props {
  registerEmail: UseFormRegisterReturn<'email'>;
  registerPassword: UseFormRegisterReturn<'password'>;
}

const LoginFormView = ({ registerEmail, registerPassword }: Props) => {
  return (
    <div className="h-full bg-gradient-to-tl from-green-400 to-indigo-900 w-full py-16 px-4">
      <div className="flex flex-col items-center justify-center">
        <svg width={188} height={74} viewBox="0 0 188 74" fill="none" xmlns="http://www.w3.org/2000/svg">
          <path
            d="M21.6477 59.0909C21.6477 52.0384 17.3438 47.8835 11.6229 47.8835C5.89134 47.8835 1.59801 52.0384 1.59801 59.0909C1.59801 66.1328 5.89134 70.2983 11.6229 70.2983C17.3438 70.2983 21.6477 66.1435 21.6477 59.0909ZM17.674 59.0909C17.674 64.0554 15.1811 66.7507 11.6229 66.7507C8.07528 66.7507 5.57173 64.0554 5.57173 59.0909C5.57173 54.1264 8.07528 51.4311 11.6229 51.4311C15.1811 51.4311 17.674 54.1264 17.674 59.0909ZM39.8304 53.6364H35.7289L31.9576 65.7919H31.7871L28.0265 53.6364H23.9142L29.7417 70H34.003L39.8304 53.6364ZM49.0803 70.3196C53.8743 70.3196 56.9212 66.9425 56.9212 61.8821C56.9212 56.8111 53.8743 53.4233 49.0803 53.4233C44.2862 53.4233 41.2393 56.8111 41.2393 61.8821C41.2393 66.9425 44.2862 70.3196 49.0803 70.3196ZM49.1016 67.2301C46.4489 67.2301 45.1491 64.8651 45.1491 61.8714C45.1491 58.8778 46.4489 56.4808 49.1016 56.4808C51.7116 56.4808 53.0114 58.8778 53.0114 61.8714C53.0114 64.8651 51.7116 67.2301 49.1016 67.2301ZM64.051 60.4119C64.051 58.0469 65.4785 56.6832 67.5133 56.6832C69.5055 56.6832 70.6987 57.9936 70.6987 60.1776V70H74.5552V59.581C74.5659 55.6605 72.3287 53.4233 68.9515 53.4233C66.5012 53.4233 64.818 54.5952 64.0723 56.4169H63.8805V53.6364H60.1944V70H64.051V60.4119ZM78.6461 70H82.5985V62.0099H86.6255L90.9082 70H95.3187L90.5993 61.3494C93.1667 60.3161 94.5197 58.1214 94.5197 55.1598C94.5197 50.9943 91.835 48.1818 86.8279 48.1818H78.6461V70ZM82.5985 58.7287V51.4844H86.2207C89.1717 51.4844 90.4927 52.8374 90.4927 55.1598C90.4927 57.4822 89.1717 58.7287 86.242 58.7287H82.5985ZM108.417 63.1179C108.417 65.6108 106.638 66.8466 104.933 66.8466C103.08 66.8466 101.844 65.5362 101.844 63.4588V53.6364H97.9874V64.0554C97.9874 67.9865 100.225 70.2131 103.442 70.2131C105.892 70.2131 107.618 68.924 108.364 67.0916H108.534V70H112.274V53.6364H108.417V63.1179ZM123.484 70.3196C127.298 70.3196 129.918 68.4553 130.6 65.6108L126.999 65.206C126.477 66.5909 125.199 67.3153 123.537 67.3153C121.044 67.3153 119.393 65.6747 119.361 62.8729H130.76V61.6903C130.76 55.9482 127.308 53.4233 123.281 53.4233C118.594 53.4233 115.536 56.8643 115.536 61.9141C115.536 67.049 118.551 70.3196 123.484 70.3196ZM119.371 60.2734C119.489 58.1854 121.033 56.4276 123.335 56.4276C125.55 56.4276 127.042 58.0469 127.063 60.2734H119.371ZM140.095 70.2876C142.907 70.2876 144.175 68.6151 144.782 67.4219H145.017V70H148.809V48.1818H144.942V56.3423H144.782C144.197 55.1598 142.993 53.4233 140.106 53.4233C136.324 53.4233 133.351 56.3849 133.351 61.8395C133.351 67.2301 136.238 70.2876 140.095 70.2876ZM141.171 67.1236C138.625 67.1236 137.282 64.8864 137.282 61.8182C137.282 58.7713 138.604 56.5874 141.171 56.5874C143.653 56.5874 145.017 58.6435 145.017 61.8182C145.017 64.9929 143.632 67.1236 141.171 67.1236ZM160.163 70.3196C163.977 70.3196 166.598 68.4553 167.28 65.6108L163.679 65.206C163.157 66.5909 161.879 67.3153 160.217 67.3153C157.724 67.3153 156.072 65.6747 156.04 62.8729H167.44V61.6903C167.44 55.9482 163.988 53.4233 159.961 53.4233C155.273 53.4233 152.216 56.8643 152.216 61.9141C152.216 67.049 155.231 70.3196 160.163 70.3196ZM156.051 60.2734C156.168 58.1854 157.713 56.4276 160.014 56.4276C162.23 56.4276 163.722 58.0469 163.743 60.2734H156.051ZM174.559 60.4119C174.559 58.0469 175.986 56.6832 178.021 56.6832C180.013 56.6832 181.206 57.9936 181.206 60.1776V70H185.063V59.581C185.074 55.6605 182.836 53.4233 179.459 53.4233C177.009 53.4233 175.326 54.5952 174.58 56.4169H174.388V53.6364H170.702V70H174.559V60.4119Z"
            fill="white"
          />
          <path
            d="M69 17.0551C69.0331 25.2688 75.7248 32 83.9453 32C92.1861 32 98.8902 25.2959 98.8902 17.0551V14.9453C98.8902 11.5521 101.651 8.79118 105.044 8.79118C108.438 8.79118 111.199 11.5521 111.199 14.9453C111.199 15.9163 111.986 16.7037 112.957 16.7037H118.232C119.203 16.7037 119.99 15.9163 119.99 14.9453C119.99 6.70457 113.286 0 105.045 0C96.8041 0 90.0995 6.70457 90.0995 14.9453V17.0551C90.0995 20.4489 87.3386 23.2088 83.9458 23.2088C80.5526 23.2088 77.7917 20.4489 77.7917 17.0551C77.7917 16.0842 77.0043 15.2968 76.0333 15.2968H70.7583C69.7874 15.2973 69 16.0842 69 17.0551Z"
            fill="white"
          />
        </svg>
        <div className="bg-white shadow rounded lg:w-1/3  md:w-1/2 w-full p-10 mt-16">
          <p
            role="heading"
            aria-level={1}
            aria-label="Login to your account"
            className="text-2xl font-extrabold leading-6 text-gray-800"
          >
            Login to your account
          </p>
          <p className="text-sm mt-4 font-medium leading-none text-gray-500">
            Dont have account?{' '}
            <span
              tabIndex={0}
              role="link"
              aria-label="Sign up here"
              className="text-sm font-medium leading-none underline text-gray-800 cursor-pointer"
            >
              {' '}
              Sign up here
            </span>
          </p>
          <button
            type="button"
            aria-label="Continue with google"
            className="focus:outline-none focus:ring-2 focus:ring-offset-1 focus:ring-gray-700 py-3.5 px-4 border rounded-lg border-gray-700 flex items-center w-full mt-10"
          >
            <svg width={19} height={20} viewBox="0 0 19 20" fill="none" xmlns="http://www.w3.org/2000/svg">
              <path
                d="M18.9892 10.1871C18.9892 9.36767 18.9246 8.76973 18.7847 8.14966H9.68848V11.848H15.0277C14.9201 12.767 14.3388 14.1512 13.047 15.0812L13.0289 15.205L15.905 17.4969L16.1042 17.5173C17.9342 15.7789 18.9892 13.221 18.9892 10.1871Z"
                fill="#4285F4"
              />
              <path
                d="M9.68813 19.9314C12.3039 19.9314 14.4999 19.0455 16.1039 17.5174L13.0467 15.0813C12.2286 15.6682 11.1306 16.0779 9.68813 16.0779C7.12612 16.0779 4.95165 14.3395 4.17651 11.9366L4.06289 11.9465L1.07231 14.3273L1.0332 14.4391C2.62638 17.6946 5.89889 19.9314 9.68813 19.9314Z"
                fill="#34A853"
              />
              <path
                d="M4.17667 11.9366C3.97215 11.3165 3.85378 10.6521 3.85378 9.96562C3.85378 9.27905 3.97215 8.6147 4.16591 7.99463L4.1605 7.86257L1.13246 5.44363L1.03339 5.49211C0.37677 6.84302 0 8.36005 0 9.96562C0 11.5712 0.37677 13.0881 1.03339 14.4391L4.17667 11.9366Z"
                fill="#FBBC05"
              />
              <path
                d="M9.68807 3.85336C11.5073 3.85336 12.7344 4.66168 13.4342 5.33718L16.1684 2.59107C14.4892 0.985496 12.3039 0 9.68807 0C5.89885 0 2.62637 2.23672 1.0332 5.49214L4.16573 7.99466C4.95162 5.59183 7.12608 3.85336 9.68807 3.85336Z"
                fill="#EB4335"
              />
            </svg>
            <p className="text-base font-medium ml-4 text-gray-700">Continue with Google</p>
          </button>
          <button
            aria-label="Continue with github"
            type="button"
            className="focus:outline-none  focus:ring-2 focus:ring-offset-1 focus:ring-gray-700 py-3.5 px-4 border rounded-lg border-gray-700 flex items-center w-full mt-4"
          >
            <svg width={21} height={20} viewBox="0 0 21 20" fill="none" xmlns="http://www.w3.org/2000/svg">
              <path
                d="M10.1543 0C4.6293 0 0.154298 4.475 0.154298 10C0.153164 12.0993 0.813112 14.1456 2.04051 15.8487C3.26792 17.5517 5.00044 18.8251 6.9923 19.488C7.4923 19.575 7.6793 19.275 7.6793 19.012C7.6793 18.775 7.6663 17.988 7.6663 17.15C5.1543 17.613 4.5043 16.538 4.3043 15.975C4.1913 15.687 3.7043 14.8 3.2793 14.562C2.9293 14.375 2.4293 13.912 3.2663 13.9C4.0543 13.887 4.6163 14.625 4.8043 14.925C5.7043 16.437 7.1423 16.012 7.7163 15.75C7.8043 15.1 8.0663 14.663 8.3543 14.413C6.1293 14.163 3.8043 13.3 3.8043 9.475C3.8043 8.387 4.1913 7.488 4.8293 6.787C4.7293 6.537 4.3793 5.512 4.9293 4.137C4.9293 4.137 5.7663 3.875 7.6793 5.163C8.49336 4.93706 9.33447 4.82334 10.1793 4.825C11.0293 4.825 11.8793 4.937 12.6793 5.162C14.5913 3.862 15.4293 4.138 15.4293 4.138C15.9793 5.513 15.6293 6.538 15.5293 6.788C16.1663 7.488 16.5543 8.375 16.5543 9.475C16.5543 13.313 14.2173 14.163 11.9923 14.413C12.3543 14.725 12.6673 15.325 12.6673 16.263C12.6673 17.6 12.6543 18.675 12.6543 19.013C12.6543 19.275 12.8423 19.587 13.3423 19.487C15.3273 18.8168 17.0522 17.541 18.2742 15.8392C19.4962 14.1373 20.1537 12.0951 20.1543 10C20.1543 4.475 15.6793 0 10.1543 0Z"
                fill="#333333"
              />
            </svg>
            <p className="text-base font-medium ml-4 text-gray-700">Continue with Github</p>
          </button>
          <button
            aria-label="Continue with twitter"
            type="button"
            className="focus:outline-none  focus:ring-2 focus:ring-offset-1 focus:ring-gray-700 py-3.5 px-4 border rounded-lg border-gray-700 flex items-center w-full mt-4"
          >
            <svg width={24} height={24} viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
              <path
                d="M22.1623 5.656C21.3989 5.9937 20.5893 6.21548 19.7603 6.314C20.634 5.79144 21.288 4.96902 21.6003 4C20.7803 4.488 19.8813 4.83 18.9443 5.015C18.3149 4.34158 17.4807 3.89497 16.5713 3.74459C15.6618 3.59421 14.7282 3.74849 13.9156 4.18346C13.1029 4.61842 12.4567 5.30969 12.0774 6.1498C11.6981 6.9899 11.607 7.93178 11.8183 8.829C10.1554 8.74566 8.52863 8.31353 7.04358 7.56067C5.55854 6.80781 4.24842 5.75105 3.1983 4.459C2.82659 5.09745 2.63125 5.82323 2.6323 6.562C2.6323 8.012 3.3703 9.293 4.4923 10.043C3.82831 10.0221 3.17893 9.84278 2.5983 9.52V9.572C2.5985 10.5377 2.93267 11.4736 3.54414 12.2211C4.15562 12.9685 5.00678 13.4815 5.9533 13.673C5.33691 13.84 4.6906 13.8647 4.0633 13.745C4.33016 14.5762 4.8503 15.3032 5.55089 15.8241C6.25147 16.345 7.09742 16.6338 7.9703 16.65C7.10278 17.3313 6.10947 17.835 5.04718 18.1322C3.98488 18.4294 2.87442 18.5143 1.7793 18.382C3.69099 19.6114 5.91639 20.2641 8.1893 20.262C15.8823 20.262 20.0893 13.889 20.0893 8.362C20.0893 8.182 20.0843 8 20.0763 7.822C20.8952 7.23017 21.6019 6.49702 22.1633 5.657L22.1623 5.656Z"
                fill="#1DA1F2"
              />
            </svg>
            <p className="text-base font-medium ml-4 text-gray-700">Continue with Twitter</p>
          </button>
          <div className="w-full flex items-center justify-between py-5">
            <hr className="w-full bg-gray-400" />
            <p className="text-base font-medium leading-4 px-2.5 text-gray-400">OR</p>
            <hr className="w-full bg-gray-400  " />
          </div>
          <div>
            <label className="text-sm font-medium leading-none text-gray-800" htmlFor="email-input">
              Email
            </label>
            <input
              aria-label="enter email adress"
              id="email-input"
              type="email"
              className="bg-gray-200 border rounded focus:outline-none text-xs font-medium leading-none text-gray-800 py-3 w-full pl-3 mt-2"
              {...registerEmail}
            />
          </div>
          <div className="mt-6  w-full">
            <label className="text-sm font-medium leading-none text-gray-800" htmlFor="password-input">
              Password
            </label>
            <div className="relative flex items-center justify-center">
              <input
                id="password-input"
                aria-label="enter Password"
                type="password"
                className="bg-gray-200 border rounded focus:outline-none text-xs font-medium leading-none text-gray-800 py-3 w-full pl-3 mt-2"
                {...registerPassword}
              />
              <div className="absolute right-0 mt-2 mr-3 cursor-pointer">
                <svg width={16} height={16} viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg">
                  <path
                    d="M7.99978 2C11.5944 2 14.5851 4.58667 15.2124 8C14.5858 11.4133 11.5944 14 7.99978 14C4.40511 14 1.41444 11.4133 0.787109 8C1.41378 4.58667 4.40511 2 7.99978 2ZM7.99978 12.6667C9.35942 12.6664 10.6787 12.2045 11.7417 11.3568C12.8047 10.509 13.5484 9.32552 13.8511 8C13.5473 6.67554 12.8031 5.49334 11.7402 4.64668C10.6773 3.80003 9.35864 3.33902 7.99978 3.33902C6.64091 3.33902 5.32224 3.80003 4.25936 4.64668C3.19648 5.49334 2.45229 6.67554 2.14844 8C2.45117 9.32552 3.19489 10.509 4.25787 11.3568C5.32085 12.2045 6.64013 12.6664 7.99978 12.6667ZM7.99978 11C7.20413 11 6.44106 10.6839 5.87846 10.1213C5.31585 9.55871 4.99978 8.79565 4.99978 8C4.99978 7.20435 5.31585 6.44129 5.87846 5.87868C6.44106 5.31607 7.20413 5 7.99978 5C8.79543 5 9.55849 5.31607 10.1211 5.87868C10.6837 6.44129 10.9998 7.20435 10.9998 8C10.9998 8.79565 10.6837 9.55871 10.1211 10.1213C9.55849 10.6839 8.79543 11 7.99978 11ZM7.99978 9.66667C8.4418 9.66667 8.86573 9.49107 9.17829 9.17851C9.49085 8.86595 9.66644 8.44203 9.66644 8C9.66644 7.55797 9.49085 7.13405 9.17829 6.82149C8.86573 6.50893 8.4418 6.33333 7.99978 6.33333C7.55775 6.33333 7.13383 6.50893 6.82126 6.82149C6.5087 7.13405 6.33311 7.55797 6.33311 8C6.33311 8.44203 6.5087 8.86595 6.82126 9.17851C7.13383 9.49107 7.55775 9.66667 7.99978 9.66667Z"
                    fill="#71717A"
                  />
                </svg>
              </div>
            </div>
          </div>
          <div className="mt-8">
            <button
              type="submit"
              aria-label="create my account"
              className="focus:ring-2 focus:ring-offset-2 focus:ring-indigo-700 text-sm font-semibold leading-none text-white focus:outline-none bg-indigo-700 border rounded hover:bg-indigo-600 py-4 w-full"
            >
              Create my account
            </button>
          </div>
        </div>
      </div>
    </div>
  );
};

export default LoginFormView;
