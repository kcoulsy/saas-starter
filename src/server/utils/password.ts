import crypto from 'crypto';

export async function hash(password: string) {
  return new Promise<string>((resolve, reject) => {
    // generate random 16 bytes long salt
    const salt = crypto.randomBytes(16).toString('hex');

    crypto.scrypt(password, salt, 64, (err, derivedKey) => {
      if (err) reject(err);
      resolve(`${salt}:${derivedKey.toString('hex')}`);
    });
  });
}

export async function verify(password: string, hashString: string) {
  return new Promise((resolve, reject) => {
    const [salt, key] = hashString.split(':');
    // Mitigating timing attacks
    // https://dev.to/bdougherty/comment/1c1fa
    const keyBuffer = Buffer.from(key || '', 'hex');
    crypto.scrypt(password, salt || '', 64, (err, derivedKey) => {
      if (err) reject(err);
      resolve(crypto.timingSafeEqual(keyBuffer, derivedKey));
    });
  });
}
