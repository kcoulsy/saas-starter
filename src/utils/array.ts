/**
 * Helper method for array filtering to maintain type safety
 * rather than using .filter(Boolean)
 *
 * @param value unknown value
 * @returns boolean
 */
export function notEmpty<TValue>(value: TValue | null | undefined): value is TValue {
  return value !== null && value !== undefined;
}
