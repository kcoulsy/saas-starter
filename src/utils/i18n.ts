import { initReactI18next } from 'react-i18next';
import i18n from 'i18next';
import enCommonTranslations from '@src/locales/en/common.json';
import enLoginFormTranslations from '@src/locales/en/login-form.json';
import enRegisterFormTranslations from '@src/locales/en/register-form.json';

i18n.use(initReactI18next).init({
  lng: 'en',
  fallbackLng: 'en',

  // have a common namespace used around the full app
  ns: ['translations'],
  defaultNS: 'translations',

  interpolation: {
    escapeValue: false, // not needed for react!!
  },

  debug: true,

  resources: {
    en: {
      common: enCommonTranslations,
      'login-form': enLoginFormTranslations,
      'register-form': enRegisterFormTranslations,
    },
  },
});

export default i18n;
