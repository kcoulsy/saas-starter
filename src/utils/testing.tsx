import React from 'react';
import { render } from '@testing-library/react';
// import TypesafeI18n from '@src/i18n/i18n-react';
// import { loadAllLocales } from '@src/i18n/i18n-util.sync';
import Providers from '@src/app/providers';

// const wrapper = ({ children }: { children: React.ReactNode }) => {
//   // loadAllLocales();
//   return <TypesafeI18n locale="en">{children}</TypesafeI18n>;
// };

const renderComponent = (component: React.ReactElement) => {
  return render(component, { wrapper: Providers });
};

export { renderComponent };
