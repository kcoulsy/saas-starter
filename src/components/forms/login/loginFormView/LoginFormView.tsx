import { useI18nContext } from '@src/i18n/i18n-react';
import FormLayout from '@src/components/common/formLayout/FormLayout';
import type { UseFormRegisterReturn } from 'react-hook-form';

export interface LoginFormViewProps {
  registerEmail: UseFormRegisterReturn<'email'>;
  errors?: {
    email?: string[];
    loginError?: string[];
  };
  isLoggingIn?: boolean;
  loginSuccess?: boolean;
}

const LoginFormView = ({ registerEmail, errors, loginSuccess, isLoggingIn = false }: LoginFormViewProps) => {
  const { LL } = useI18nContext();

  return (
    <div className="bg-white shadow rounded w-full p-10">
      {loginSuccess ? (
        <div className="text-center">
          <h1 className="text-2xl font-semibold">{LL.login.form.emailSuccess()}</h1>
          <p className="mt-4">{LL.login.form.emailSuccessSubcopy()}</p>
        </div>
      ) : (
        <>
          {/* <div className="grid gap-y-4 mt-8">
        <SocialLoginButton type="google" onClick={() => alert('login with google')} />
        <SocialLoginButton type="twitter" onClick={() => alert('login with google')} />
        <SocialLoginButton type="github" onClick={() => alert('login with google')} />
      </div> */}

          {/* <div className="w-full flex items-center justify-between py-5">
        <hr className="w-full bg-gray-400" />
        <p className="text-base font-medium leading-4 px-2.5 text-gray-400 uppercase">
          {LL.login.form.seperatorLabel()}
        </p>
        <hr className="w-full bg-gray-400  " />
      </div> */}
          <FormLayout
            title={LL.login.form.title()}
            formFields={[
              {
                id: 'email',
                type: 'email',
                label: LL.login.form.emailLabel(),
                placeholder: LL.login.form.emailPlaceholder(),
                register: registerEmail,
                errors: errors?.email || [],
              },
            ]}
            submitButtonLabel={LL.login.form.submitButton()}
            errors={errors?.loginError || []}
            isLoading={isLoggingIn}
          />
        </>
      )}
    </div>
  );
};

export default LoginFormView;
