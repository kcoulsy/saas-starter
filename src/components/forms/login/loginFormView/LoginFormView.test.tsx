import { renderComponent } from '@src/utils/testing';
import { composeStories } from '@storybook/react';
import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe } from 'vitest';
import * as stories from './LoginFormView.stories';

const { Default, Errors } = composeStories(stories);

const mockRegisterEmail = { onBlur: vi.fn(), onChange: vi.fn(), name: 'email', ref: vi.fn() } as const;

describe('LoginFormView', () => {
  test('it renders Default and matches snapshots', () => {
    const { asFragment } = renderComponent(<Default />);

    expect(asFragment()).toMatchSnapshot();
  });

  test('it renders Errors and matches snapshots', () => {
    const { asFragment } = renderComponent(<Errors />);

    expect(asFragment()).toMatchSnapshot();
  });

  test('it renders and registers email field', async () => {
    renderComponent(<Default registerEmail={mockRegisterEmail} />);

    const emailField = screen.getByLabelText('Email');
    expect(emailField).toBeInTheDocument();

    await userEvent.type(emailField, 'something@email.com');

    expect(mockRegisterEmail.onChange).toHaveBeenCalled();
  });

  it('renders login error', () => {
    renderComponent(<Default errors={{ loginError: ['Login error test 1', 'Login error test 2'] }} />);
    const loginError1 = screen.getByText('Login error test 1');
    const loginError2 = screen.getByText('Login error test 2');
    expect(loginError1).toBeInTheDocument();
    expect(loginError2).toBeInTheDocument();
  });
});
