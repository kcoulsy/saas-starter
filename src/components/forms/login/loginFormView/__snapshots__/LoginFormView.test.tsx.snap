// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginFormView > it renders Default and matches snapshots 1`] = `
<DocumentFragment>
  <div
    class="lg:w-1/3 md:w-1/2 w-full"
  >
    <div
      class="bg-white shadow rounded w-full p-10"
    >
      <div
        class=""
      >
        <div
          class=""
        >
          <h1
            aria-label="Login or Sign up"
            aria-level="1"
            class="text-2xl font-extrabold leading-6 text-gray-800"
          >
            Login or Sign up
          </h1>
          <div
            class="mt-4 w-full grid gap-y-4 "
          >
            <div>
              <div
                class="relative flex align-middle"
              >
                <label
                  class="text-sm font-medium leading-none text-gray-800 capitalize relative flex-1"
                  for="email"
                >
                  Email
                  <input
                    aria-label="Email"
                    class="bg-gray-200 border rounded focus:outline-none text-xs font-medium leading-none text-gray-800 py-3 w-full pl-3 mt-2"
                    id="email"
                    name="email"
                    placeholder="Enter your email"
                    type="email"
                  />
                </label>
              </div>
              <ul
                class="mt-1"
              />
            </div>
          </div>
          <ul />
          <button
            aria-label="Login or Sign up"
            class="text-sm font-semibold leading-none text-white focus:outline-none flex justify-center items-center h-10 w-full bg-indigo-700 border rounded hover:bg-indigo-600 focus:ring-2 focus:ring-offset-2 focus:ring-indigo-700 mt-4"
            type="submit"
          >
            Login or Sign up
          </button>
        </div>
      </div>
    </div>
  </div>
  ;
</DocumentFragment>
`;

exports[`LoginFormView > it renders Errors and matches snapshots 1`] = `
<DocumentFragment>
  <div
    class="lg:w-1/3 md:w-1/2 w-full"
  >
    <div
      class="bg-white shadow rounded w-full p-10"
    >
      <div
        class=""
      >
        <div
          class=""
        >
          <h1
            aria-label="Login or Sign up"
            aria-level="1"
            class="text-2xl font-extrabold leading-6 text-gray-800"
          >
            Login or Sign up
          </h1>
          <div
            class="mt-4 w-full grid gap-y-4 "
          >
            <div>
              <div
                class="relative flex align-middle"
              >
                <label
                  class="text-sm font-medium leading-none text-gray-800 capitalize relative flex-1"
                  for="email"
                >
                  Email
                  <input
                    aria-label="Email"
                    class="bg-gray-200 border rounded focus:outline-none text-xs font-medium leading-none text-gray-800 py-3 w-full pl-3 mt-2"
                    id="email"
                    name="email"
                    placeholder="Enter your email"
                    type="email"
                  />
                </label>
              </div>
              <ul
                class="mt-1"
              >
                <li>
                  <small
                    class="text-red-500"
                  >
                    Please enter your email
                  </small>
                </li>
              </ul>
            </div>
          </div>
          <ul>
            <li
              class="mt-1"
            >
              <small
                class="text-red-500"
              >
                Invalid email or password
              </small>
            </li>
            <li
              class="mt-1"
            >
              <small
                class="text-red-500"
              >
                Failed to reach server
              </small>
            </li>
          </ul>
          <button
            aria-label="Login or Sign up"
            class="text-sm font-semibold leading-none text-white focus:outline-none flex justify-center items-center h-10 w-full bg-indigo-700 border rounded hover:bg-indigo-600 focus:ring-2 focus:ring-offset-2 focus:ring-indigo-700 mt-4"
            type="submit"
          >
            Login or Sign up
          </button>
        </div>
      </div>
    </div>
  </div>
  ;
</DocumentFragment>
`;
